bitcoin/src/bip32.rs

```diff
@@ -593,7 +593,11 @@ impl Xpriv {
             depth: 0,
             parent_fingerprint: Default::default(),
             child_number: ChildNumber::ZERO_NORMAL,
-            private_key: secp256k1::SecretKey::from_slice(&hmac_result.as_ref()[..32])?,
+            private_key: secp256k1::SecretKey::from_byte_array(
+                &hmac_result.as_ref()[..32]
+                    .try_into()
+                    .expect("Internal HMAC hash is broken: shorter than 32 bytes"),
+            )?,
             chain_code: ChainCode::from_hmac(hmac_result),
```

Comment:
Perhaps it would be nicer to write this using a local variable above

```rust
        let mut key = [0_u8; 32];
        key.copy_from_slice(&hmac_result.as_ref()[..32]);
```

Then use 
```rust
private_key: secp256k1::SecretKey::from_byte_array(&key)?,
```

Same for all other changes to `from_slice`.

Reply:
I prefer `try_into` which is explicit, less error-prone and more performant.

That being said, the expect message should say "length statically known" because the `[..32]` operator forces it. (If it panics, the operator would have to be broken, not hash.)

Reply:
It's an unfortunate choice between `mut` and `expect`. I slightly lean toward Kix that the `expect` is less bad.

Reply:
Note that `copy_from_slice` still has a panic branch in it, it's just not clearly visible from the name. So there is no perf difference (and anyway I'm sure that in both cases the branch will be optimized away).

Reply:
Works for me, thanks.